File: apps/api/app/services/scoring.py
"""Deterministic product opportunity scoring.

Agents call this as a tool instead of inventing scores, so launch decisions
are reproducible and auditable. Weights follow the product strategy:
demand 30%, competition 20%, margin 25%, trend 15%, risk 10%.

All factor inputs are 0-100. `competition` and `risk` are supplied as raw
levels (higher = worse) and inverted internally.
"""

from dataclasses import asdict, dataclass

from app.core.config import get_settings

WEIGHTS = {
    "demand": 0.30,
    "competition": 0.20,
    "margin": 0.25,
    "trend": 0.15,
    "risk": 0.10,
}

VERDICT_LAUNCH = "launch"
VERDICT_WATCH = "watch"
VERDICT_REJECT = "reject"

WATCH_THRESHOLD = 70.0


@dataclass
class ProductScore:
    demand: float
    competition: float
    margin: float
    trend: float
    risk: float
    total_score: float
    verdict: str
    rationale: str

    def as_dict(self) -> dict:
        return asdict(self)


def _clamp(value: float) -> float:
    return max(0.0, min(100.0, float(value)))


def margin_score_from_prices(cost: float, price: float) -> float:
    """Map a gross margin percentage onto a 0-100 score.

    Dropshipping viability roughly requires >= 60% gross margin for a top
    score; 30% is breakeven territory once ads are factored in.
    """
    if price <= 0:
        return 0.0
    margin_pct = (price - max(cost, 0.0)) / price * 100
    return _clamp((margin_pct / 60.0) * 100)


def score_product(
    demand: float,
    competition: float,
    margin: float,
    trend: float,
    risk: float,
) -> ProductScore:
    demand = _clamp(demand)
    competition = _clamp(competition)
    margin = _clamp(margin)
    trend = _clamp(trend)
    risk = _clamp(risk)

    total = (
        demand * WEIGHTS["demand"]
        + (100 - competition) * WEIGHTS["competition"]
        + margin * WEIGHTS["margin"]
        + trend * WEIGHTS["trend"]
        + (100 - risk) * WEIGHTS["risk"]
    )
    total = round(total, 2)

    threshold = get_settings().launch_score_threshold
    if total >= threshold:
        verdict = VERDICT_LAUNCH
        rationale = f"Score {total} meets the launch threshold of {threshold}."
    elif total >= WATCH_THRESHOLD:
        verdict = VERDICT_WATCH
        rationale = (
            f"Score {total} is promising but below the launch threshold of {threshold}; "
            "monitor and re-evaluate."
        )
    else:
        verdict = VERDICT_REJECT
        rationale = f"Score {total} is below the watch threshold of {WATCH_THRESHOLD}."

    return ProductScore(
        demand=demand,
        competition=competition,
        margin=margin,
        trend=trend,
        risk=risk,
        total_score=total,
        verdict=verdict,
        rationale=rationale,
    )

File: apps/web/lib/scoring.ts
/**
 * Deterministic product opportunity scoring — ported verbatim from the API.
 * Agents call this as a tool instead of inventing scores, so launch decisions
 * are reproducible and auditable.
 *
 * Weights: demand 30%, competition 20%, margin 25%, trend 15%, risk 10%.
 * All inputs are 0-100. `competition` and `risk` are raw levels (higher = worse)
 * and are inverted internally.
 */
import { LAUNCH_SCORE_THRESHOLD } from "./config";

export const WEIGHTS = {
  demand: 0.3,
  competition: 0.2,
  margin: 0.25,
  trend: 0.15,
  risk: 0.1,
} as const;

export const VERDICT_LAUNCH = "launch";
export const VERDICT_WATCH = "watch";
export const VERDICT_REJECT = "reject";
export const WATCH_THRESHOLD = 70.0;

export interface ProductScore {
  demand: number;
  competition: number;
  margin: number;
  trend: number;
  risk: number;
  total_score: number;
  verdict: string;
  rationale: string;
}

const clamp = (v: number) => Math.max(0, Math.min(100, Number(v) || 0));

export function marginScoreFromPrices(cost: number, price: number): number {
  if (price <= 0) return 0;
  const marginPct = ((price - Math.max(cost, 0)) / price) * 100;
  return clamp((marginPct / 60) * 100);
}

export function scoreProduct(input: {
  demand: number;
  competition: number;
  margin: number;
  trend: number;
  risk: number;
}): ProductScore {
  const demand = clamp(input.demand);
  const competition = clamp(input.competition);
  const margin = clamp(input.margin);
  const trend = clamp(input.trend);
  const risk = clamp(input.risk);

  const total =
    Math.round(
      (demand * WEIGHTS.demand +
        (100 - competition) * WEIGHTS.competition +
        margin * WEIGHTS.margin +
        trend * WEIGHTS.trend +
        (100 - risk) * WEIGHTS.risk) *
        100,
    ) / 100;

  let verdict: string;
  let rationale: string;
  if (total >= LAUNCH_SCORE_THRESHOLD) {
    verdict = VERDICT_LAUNCH;
    rationale = `Score ${total} meets the launch threshold of ${LAUNCH_SCORE_THRESHOLD}.`;
  } else if (total >= WATCH_THRESHOLD) {
    verdict = VERDICT_WATCH;
    rationale = `Score ${total} is promising but below the launch threshold of ${LAUNCH_SCORE_THRESHOLD}; monitor and re-evaluate.`;
  } else {
    verdict = VERDICT_REJECT;
    rationale = `Score ${total} is below the watch threshold of ${WATCH_THRESHOLD}.`;
  }

  return { demand, competition, margin, trend, risk, total_score: total, verdict, rationale };
}
